File: observalog-node/src/emit.ts
import { F, Err, Outcome, Level } from './fields';
import { getContext, LogContextData } from './context';
import { applyGracefulDegradation } from './degrade';
import { encodePartA, encodePartB, LogEntry } from './wire';
import { send } from './async-drain';
import { SERVICE_SYSTEM, SERVICE_AUTH, SERVICE_DOC, SERVICE_PROVIDER } from './wire-contract';

let minLevel    = Level.Info;
let serviceCode = SERVICE_SYSTEM;

export function setMinLevel(l: Level): void    { minLevel = l; }
export function setServiceCode(c: number): void { serviceCode = c; }
export function getServiceCode(): number        { return serviceCode; }

export function debug(event: string, message: string, fields: F = {}): void {
    emitLine(Level.Debug, event, message, fields);
}

export function info(event: string, message: string, fields: F = {}): void {
    emitLine(Level.Info, event, message, fields);
}

export function warn(event: string, message: string, fields: F = {}): void {
    emitLine(Level.Warn, event, message, fields);
}

export function error(event: string, message: string, fields: F = {}): void {
    emitLine(Level.Error, event, message, fields);
}

function isErrLike(v: unknown): v is Err | { kind: string; code: string; message: string; retryable: boolean } {
    return v !== null && typeof v === 'object' && 'code' in v && 'message' in v;
}

function emitLine(level: Level, event: string, message: string, fields: F): void {
    if (level < minLevel) return;

    // Get or create a degraded context
    const stored = getContext();
    const ctx: LogContextData = stored ?? {
        traceId: '', spanId: '', parentSpan: '',
        userId: '', journeyStage: '', seq: { value: 0 },
    };
    applyGracefulDegradation(ctx);

    // Extract structural fields from F
    let outcome:    Outcome | undefined;
    let durationMs: number  | undefined;
    let errField:   Err     | undefined;
    const devCtx: Record<string, unknown> = {};

    for (const [k, v] of Object.entries(fields)) {
        if (k === 'outcome' && typeof v === 'string') {
            outcome = v as Outcome;
        } else if (k === 'duration_ms' && typeof v === 'number') {
            durationMs = v;
        } else if (k === 'error' && isErrLike(v)) {
            // Accept both `new Err(...)` instances and plain objects with the same shape
            errField = v instanceof Err
                ? v
                : new Err(
                    String((v as any).kind    ?? ''),
                    String((v as any).code    ?? ''),
                    String((v as any).message ?? ''),
                    Boolean((v as any).retryable),
                  );
        } else {
            devCtx[k] = v;
        }
    }

    const outcomeCode = outcomeToCode(outcome);
    const seqVal = ++ctx.seq.value;

    const entry: LogEntry = {
        traceId:      ctx.traceId,
        spanId:       ctx.spanId,
        parentSpan:   ctx.parentSpan,
        seq:          seqVal,
        serviceCode,
        levelCode:    level,
        outcomeCode,
        tsMs:         Date.now(),
        event,
        message,
        durationMs,
        outcome,
        error:        errField,
        userId:       ctx.userId,
        journeyStage: ctx.journeyStage,
        ctx:          devCtx,
    };

    send(encodePartA(entry));
    send(encodePartB(entry));
}

function outcomeToCode(o: Outcome | undefined): number {
    switch (o) {
        case Outcome.Success: return 1;
        case Outcome.Failure: return 2;
        case Outcome.Partial: return 3;
        case Outcome.Pending: return 4;
        default:              return 0;
    }
}

export function resolveServiceCode(name: string): number {
    switch (name) {
        case 'auth':     return SERVICE_AUTH;
        case 'doc':      return SERVICE_DOC;
        case 'provider': return SERVICE_PROVIDER;
        default:         return SERVICE_SYSTEM;
    }
}

File: observalog-go/emit.go
package log

import (
	"context"
	"sync/atomic"
	"time"
)

var (
	minLevel   Level = LevelInfo
	serviceCode uint8
)

// GetServiceCode returns the service code set at Init().
func GetServiceCode() uint8 {
	return serviceCode
}

func Info(ctx context.Context, event, message string, fields F) {
	emitLine(ctx, LevelInfo, event, message, fields)
}

func Warn(ctx context.Context, event, message string, fields F) {
	emitLine(ctx, LevelWarn, event, message, fields)
}

func Error(ctx context.Context, event, message string, fields F) {
	emitLine(ctx, LevelError, event, message, fields)
}

func Debug(ctx context.Context, event, message string, fields F) {
	emitLine(ctx, LevelDebug, event, message, fields)
}

func emitLine(ctx context.Context, level Level, event, message string, fields F) {
	// Level check: return if below threshold.
	if level < minLevel {
		return
	}

	// Apply graceful degradation for missing ambient fields.
	ctx = ApplyGracefulDegradation(ctx)

	// Extract ambient fields from context.
	traceID := TraceIDFromCtx(ctx)
	spanID := SpanIDFromCtx(ctx)
	parentSpan := ParentSpanFromCtx(ctx)
	userID := UserIDFromCtx(ctx)
	journeyStage := JourneyStageFromCtx(ctx)

	// Increment seq.
	seq := SeqFromCtx(ctx)
	if seq == nil {
		seq = new(atomic.Uint32)
		ctx = WithSeq(ctx, seq)
	}
	seqVal := seq.Add(1)

	// Extract and remove structural fields from F.
	outcome := Outcome("")
	durationMs := (*int64)(nil)
	var errField *Err
	developerCtx := make(F, len(fields))

	for k, v := range fields {
		switch k {
		case "outcome":
			if o, ok := v.(Outcome); ok {
				outcome = o
			} else if s, ok := v.(string); ok {
				outcome = Outcome(s)
			}
		case "duration_ms":
			if ms, ok := v.(int64); ok {
				durationMs = &ms
			}
		case "error":
			if e, ok := v.(*Err); ok {
				errField = e
			}
		default:
			developerCtx[k] = v
		}
	}

	// Compute fingerprint if error present.
	var fingerprint int64
	if errField != nil {
		// Extract ctx_primary_key from developer context if available.
		ctxKey := ""
		if pk, ok := developerCtx["ctx_primary_key"].(string); ok {
			ctxKey = pk
		}
		fingerprint = ComputeFingerprint(GetServiceCode(), event, errField.Code, ctxKey)
	}

	// Convert level to wire code.
	levelCode := uint8(level)

	// Convert outcome to wire code.
	outcomeCode := uint8(OutcomeCodeNone)
	switch outcome {
	case Success:
		outcomeCode = OutcomeCodeSuccess
	case Failure:
		outcomeCode = OutcomeCodeFailure
	case Partial:
		outcomeCode = OutcomeCodePartial
	case Pending:
		outcomeCode = OutcomeCodePending
	}

	// Create log entry.
	entry := LogEntry{
		TraceID:      traceID,
		SpanID:       spanID,
		ParentSpan:   parentSpan,
		Seq:          seqVal,
		ServiceCode:  GetServiceCode(),
		LevelCode:    levelCode,
		OutcomeCode:  outcomeCode,
		TsMs:         time.Now().UnixMilli(),
		Event:        event,
		Message:      message,
		DurationMs:   durationMs,
		Outcome:      outcome,
		Error:        errField,
		Fingerprint:  fingerprint,
		UserID:       userID,
		JourneyStage: journeyStage,
		Ctx:          developerCtx,
	}

	// Encode and send.
	partA := EncodePartA(entry)
	partB := EncodePartB(entry)
	SendToChannel(partA)
	SendToChannel(partB)
}
